geo/src/algorithm/vector_ops.rs

```diff
@@ -0,0 +1,347 @@
+//! This module defines the [Vector2DOps] trait and implements it for the
+//! [Coord] struct.
+//!
+//! In the future [Vector2DOps] might be implemented on other types:
+//!
+//! - Based on community discussions it seems like the existing struct
+//!   [crate::Coord] is just one of many future data structures which may hold
+//!   coordinate information. For example future data structures may also hold Z
+//!   and/or M ordinates, or other arbitrary data.
+//! - [geotraits::CoordTrait] is a future trait defining accessor methods
+//!   `.x()` and `.y()` which will facilitate these generic data structures.
+//!
+//! > Note: [Vector2DOps] is not implemented for [crate::Point] because users
+//! > probably expect to see more familiar geospatial functions like
+//! > `a.euclidean_distance(b)` at that level and generally not linear algebra
+//! > like `a.dot(b)`.
+//!
+//! For now, it is assumed that users of [Vector2DOps] are satisfied to upcast
+//! everything to [CoordFloat]. In future it might make sense to split this
+//! trait into 3 flavors supporting progressively more functions:
+//!
+//! - `trait Vector2DOpsCoordNum` for [CoordNum]
+//!     - Supports magnitude_squared, dot
+//! - `trait Vector2DOpsCoordNumSigned:Vector2DOpsCoordNum` for [CoordNum] + [Signed]
+//!     - Supports left, right, wedge_product
+//! - `trait Vector2DOpsCoordFloat:Vector2DOpsCoordNumSigned` for [CoordFloat]
+//!     - Supports magnitude, normalize, etc
+//!
+//! Maybe if these traits were restricted to the future [geotraits::CoordTrait]
+//! then they could provide default implementations using the accessors `.x()`
+//! and `.y()`?
+//!
+
+use crate::{Coord, CoordFloat, CoordNum};
+
+/// Defines vector operations for 2D coordinate types which implement CoordFloat
+///
+/// This trait is intended for internal use within the geo crate as a way to
+/// bring together the various hand-crafted linear algebra operations used
+/// throughout other algorithms and attached to various structs.
+///
+///
+
+pub trait Vector2DOps<Rhs = Self>
+where
+    Self: Sized,
+{
+    type NumericType: CoordNum + Send + Sync;

```

Comment:
```suggestion
    type Scalar: CoordNum + Send + Sync;
```

We have, I think, similar associated types elsewhere in the codebase which we call `Scalar` - can we use that naming here too to stay conventional, or is that not accurate in this case?

Reply:
Changed to `Scalar`, that is the correct terminology 👍